server/implementation/src/main/java/io/smallrye/graphql/bootstrap/FederationDataFetcher.java

```diff
@@ -0,0 +1,91 @@
+package io.smallrye.graphql.bootstrap;
+
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toSet;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import com.apollographql.federation.graphqljava._Entity;
+
+import graphql.schema.DataFetcher;
+import graphql.schema.DataFetchingEnvironment;
+import graphql.schema.DelegatingDataFetchingEnvironment;
+import graphql.schema.GraphQLArgument;
+import graphql.schema.GraphQLCodeRegistry;
+import graphql.schema.GraphQLFieldDefinition;
+import graphql.schema.GraphQLObjectType;
+import graphql.schema.GraphQLOutputType;
+
+class FederationDataFetcher implements DataFetcher<List<Object>> {
+
+    private final GraphQLObjectType queryType;
+    private final GraphQLCodeRegistry codeRegistry;
+
+    public FederationDataFetcher(GraphQLObjectType queryType, GraphQLCodeRegistry codeRegistry) {
+        this.queryType = queryType;
+        this.codeRegistry = codeRegistry;
+    }
+
+    @Override
+    public List<Object> get(DataFetchingEnvironment environment) throws Exception {
+        return environment.<List<Map<String, Object>>> getArgument(_Entity.argumentName).stream()
+                .map(representations -> fetchEntities(environment, representations))
+                .collect(toList());
+    }
+
+    private Object fetchEntities(DataFetchingEnvironment env, Map<String, Object> representations) {
+        Map<String, Object> requestedArgs = new HashMap<>(representations);
+        requestedArgs.remove("__typename");
+        String typename = (String) representations.get("__typename");
+        for (GraphQLFieldDefinition field : queryType.getFields()) {
+            if (matchesReturnType(field, typename) && matchesArguments(requestedArgs, field)) {
+                return execute(field, env, requestedArgs);
+            }
+        }
+        throw new RuntimeException("no query found for " + typename + " by " + requestedArgs.keySet());
+    }
+
+    private boolean matchesReturnType(GraphQLFieldDefinition field, String typename) {
+        GraphQLOutputType returnType = field.getType();
+        return returnType instanceof GraphQLObjectType && ((GraphQLObjectType) returnType).getName().equals(typename);
```

Comment:
In our current schema we are also use GraphQLInterfaceType for Federation return.

I'm not sure but I think this could work:

```
return (returnType instanceof GraphQLObjectType && ((GraphQLObjectType) returnType).getName().equals(typename)) || 
(returnType instanceof GraphQLInterfaceType && ((GraphQLInterfaceType) returnType).getName().equals(typename))
```

Reply:
I just pushed a change using `GraphQLNamedSchemaElement` instead, which covers `GraphQLInterfaceType` and in the future Unions, etc.